### DISCUSSION TITLE
Executor Plugin and State Management

### DISCUSSION THREAD

**Discussion by greenpau** [argoproj/argo-workflows#12026] [Q&A] [answered]
I am working on a Workflows Executor Plugin and trying figure out how to keep state for long-running task.

When I start the task, the plugin should get execution id from backend system and use that id in all subsequent invocations.

The plugin returns Pending with retry in 1 minute.

The expectation is that the subsequent call to the plugin will be passed with previously received execution id.

Reviewed a number of plugins, but did not see any code where the plugin returns some kind of id, which in turn gets passed after subsequent invocation until completion (success/failure).

**Comment by greenpau** [comment]
What is unclear whether a task gets its own dedicated “plugin” container. If that is true, and the container is not shared between tasks (workflows) then the state can be maintained globally for the instance of a plugin.

**Reply by agilgur5** [reply, marked as answer]
Generally, each task gets its own Pod and an executor plugin is [a sidecar container of the main container](https://argo-workflows.readthedocs.io/en/latest/executor_plugins/#a-simple-python-plugin). 

Every Workflow and task node has identifiers, so if you need a unique identifier, I would use the existing ones in the spec.

**Reply by agilgur5** [reply]
My knowledge of the [Agent](https://argo-workflows.readthedocs.io/en/release-3.4/http-template/#argo-agent) was incomplete when I wrote the comment above, so I wanted to make some small corrections here. The Agent is a variation of the Executor that runs for Executor Plugins and HTTP templates, and the plugins are a sidecar of the Agent specifically.
The Agent isn't particularly documented well currently, but one Agent Pod is run per Workflow that uses it, rather than per task. See also https://github.com/argoproj/argo-workflows/issues/7891 and https://github.com/argoproj/argo-workflows/discussions/7144#discussioncomment-8220779

The suggestion on using existing identifiers from the `metadata` (`uid`) and `status` (node ids) remains the same though.
